### ISSUE TITLE
npm program die at the second "Waiting for installers"

### ISSUE THREAD

**Issue by bigbenz8** [felixrieseberg/windows-build-tools#70]
I'v waited 5 hours for install fininsh !

**Comment by Tatsujinichi** [comment]
I have seen this as well.

**Comment by sparcut** [comment]
Was also waiting.

Had a look at what it downloads, in this file https://github.com/felixrieseberg/windows-build-tools/blob/6b35533f83f8ee686bec2e0896d54cd0c70b94f3/src/constants.js#L5

I went to the url which downloaded the installer, then the npm install instantly finished. So try going to https://download.microsoft.com/download/5/f/7/5f7acaeb-8363-451f-9425-68a90f98b238/visualcppbuildtools_full.exe npm installing to see if it installs.

**Comment by MuhamadOmr** [comment]
@bigbenz8 will it eventually complete the install?  .. i have been waiting for 2 hours now

**Comment by sparcut** [comment]
@MuhamadOmr Have you tried what I suggested above?

**Comment by hellokatili** [comment]
@sparcut I tried it and it didn't work. "Waiting for installers" still hangs.

**Comment by SvenSchoene** [comment]
I found the "fix" the following way: 

- I downloaded the build tools from the URL @sparcut provided.
- I tried to run the installer.
- The installer gave me the error: "Cannot run installer because another setup is in process."

It seemed obvious that the same issue was preventing progress from the installer.
I simply closed any and all applications running on my computer, tried to run the installer that was downloaded, and since that worked, I guessed that `npm install --global --production windows-build-tools` would work now as well. And indeed it did.

Hope that works for some of you.

**Comment by LemuelCushing** [comment]
@SvenSchoene's fix didn't work for me, unfortunately. The installer from @sparcut URL installed correctly, it seems, while the npm installation was still running, but complained in the end that some parts were already installed so were skipped.
Killed the windows-build-tools, started them again, and it once again hangs in the second installer.

**Comment by sparcut** [comment]
@hellokatili Have you had any success with what fixed it for @SvenSchoene?

Tracked down where the download is occurring from the `constants.js`, funnily enough, it happens in `download.js`... probably should have been the first place I looked, but anyway. https://github.com/felixrieseberg/windows-build-tools/blob/ec5bc83bc4e801dd75a98ad932686a7ee64a5647/src/download.js 

It uses npm [`nugget`](https://www.npmjs.com/package/nugget), though I've never used it before so I don't know if it's being used correctly, it does say it is a CLI installer, but it seems to be importable as a node module. I reckon the problem will either be that nugget is having trouble with installing windows build tool , or nugget is being used wrong/download.js error.

Hmmm, strange.

**Comment by Sadi-1992** [comment]
none of this solutions is working for me ...  
._.

**Comment by Begi** [comment]
Open your tasks manager then kill all processus "Microsoft Visual C++ Build Tools" then try to run the npm command or download visualcppbuildtools_full.exe and install it

**Comment by rajitha-bandara** [comment]
While waiting nearly 2 hours, it completed with the success messsage

**Comment by TobiWo** [comment]
For all people who will face this issue too, just check the solution by scriptnull in the following [post](https://github.com/felixrieseberg/windows-build-tools/issues/47). This made my day!!!

**Comment by pmacn** [comment]
When running in debug mode I noticed the following message  
```
windows-build-tools Installer: Stdout from launch-installer.ps1:
This installation package could not be opened.  Verify that the package exists and that you can access it, or contact the application vendor to verify that this is a valid Windows Installer package.
```
It looks as if it downloaded python-2.7.14.amd64.msi but then tries to run python-2.7.13.msi

```
[>                                            ] 0.0% (0 B/s)
Downloading python-2.7.14.amd64.msi
```

```
function runPythonInstaller
{
    if (Test-Path $path)
    {
        cd $path
        $pyParams = "/i", "python-2.7.13.msi", "TARGETDIR=```"$path\python27```"", "ALLUSERS=0", "/qn", "/L*P", "`"$path\python-log.txt`""
        Invoke-Expression "msiexec.exe $pyParams"
    }
}
```

**Comment by andi-georg** [comment]
@pmacn: Thanks a lot. I copied the downloaded python-2.7.14.amd64.msi to a new file with name python-2.7.13.msi, then started the installation again and it worked!!!!!

**Comment by kmiklas** [comment]
what andi-georg said. Just be careful: not only do you have to remove the amd64 from the filename, but also to change the version number from 2.7.14 to 2.7.13 (emphasis on 13).

Seems that the python filename/version is hard-coded.

**Comment by ganicke** [comment]
Yep, that is the trick:
```
C:\Users\ganicke\.windows-build-tools>copy python-2.7.14.amd64.msi python-2.7.13.msi
```

**Comment by ADTC** [comment]
I see several competing solutions here. What is the one solution that's clean and most likely to work?

**Comment by jpack2015** [comment]
Left mine downloading for over 10 hours overnight and its still hung there. None of the aforementioned suggestions have worked.

**Comment by ADTC** [comment]
Yeah me too. The second one is still spinning. I tried installing Python from the renamed installer. It installed but this is still stuck. I'm probably have to find and kill the process.

What is the reason there's no fix for something so critical after so long?

**Comment by ADTC** [comment]
@jpack2015 if it has finished installing `Visual Studio Build Tools` just press Ctrl-C and exit it. Then manually install Python 2 (you can use the downloaded file in windows-build-tools).

VS Build Tools will be in `C:\Users\<<username>>\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\bin\amd64` (replace `<<username>>`) so you would want to add this path to `PATH`.

Also, I got here because of trying to install `node-sass` in which case, also hardlink the Python executable:
```
cd C:\Python27
mklink /H python2.exe python.exe
```
**But this only fixes the immediate problem. You may encounter more problems after fixing it.** *Sorry, at the moment that's all I have.*

**Comment by jpack2015** [comment]
@ADTC Yeah, I got mine to work finally by installing python myself. Funny how it didn't work the first time I tried it, but decided to work this time.

**Comment by felixrieseberg** [comment]
This should be fixed in 1.4.1 thanks to @pmacn. Sorry for the trouble!
